.ci/libfunctions.sh

```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+function perform {
+    echo "$ $1"
+    eval $1
+}
+
+function test_bottle {
+    echo '+++ :hash: Verify Package Hash - SHA-256'
+    pushd "$BUILDKITE_PIPELINE_SLUG" # same as repo name
+    perform "curl -fsSL -o './$BOTTLE' '$ROOT_URL/$BOTTLE'"
+    FILE_HASH="openssl dgst -sha256 '$BOTTLE'"
+    echo "$ $FILE_HASH"
+    BOTTLE_HASH="$(eval $FILE_HASH | tee >(cat - >&9) | awk '{print $2}')"
+    RUBY_LINE="cat '$RUBY_FILE' | grep 'sha256' | grep '$OS'"
+    echo "$ $RUBY_LINE"
+    RUBY_HASH="$(eval $RUBY_LINE | tee >(cat - >&9) | awk '{print $3}' | tr -d '"')"
+    if [[ "$BOTTLE_HASH" == "$RUBY_HASH" ]]; then
+        echo "Pass: $BOTTLE_HASH = $RUBY_HASH"
+    else
+        echo "+++ :x: Failed to Verify Bottle Hash"
+        FAIL_MSG="$BOTTLE hash $BOTTLE_HASH ≠ $RUBY_FILE::$OS hash $RUBY_HASH"'!'
+        echo "FAILURE: $FAIL_MSG"
+        [[ "$BUILDKITE" == 'true' ]] && echo "**FAILURE:** $FAIL_MSG" | buildkite-agent annotate --style 'error' --context "wrong-hash-$OS"
+        exit 2
+    fi
+    mv $BOTTLE ..
+    popd
+}
+
+function test_binary_exists {
+    WHICH_BIN="which '$BIN'"
+    echo "$ $WHICH_BIN"
+    export BUILD_ROOT="$(eval $WHICH_BIN | tee >(cat - >&9) | sed "s_/bin/${BIN}__g")"
+    if [[ -z "$BUILD_ROOT" ]]; then
+        FAIL_MSG="'$BIN' binary not found after homebrew installation on $OS_STYLIZED"'!'
+        echo "FAILURE: $FAIL_MSG"
+        [[ "$BUILDKITE" == 'true' ]] && echo "**FAILURE:** $FAIL_MSG" | tr "'" '`' | buildkite-agent annotate --style 'error' --context "no-binary-$OS"
+        exit 3
+    else
+        echo "Found '$BIN' at \"$BUILD_ROOT/bin/$BIN\"."
+    fi
+}
+
+function test_full_version_matches {
+   echo "Using git tag '$REPO_UNDER_TEST:$GIT_TAG'."
+   perform "git clone --recursive --single-branch --branch '${GIT_TAG}' '${REPO_UNDER_TEST}'"
```

Comment:
**UX** - I like how you moved the `git clone` of [eos](https://github.com/EOSIO/eos) later in the script so that we don't have to wait for it unless we are actually using it. However, this will dump hundreds or thousands of lines of nonsense into the expanded log-folding group containing the test. Can we either suppress the output, put it into a folded log group (which is difficult because we're in the middle of the test one), or move it back to the `Configuring Environment` group? I think suppressing the output is probably the best option of the three.
```bash
perform "git clone --recursive --single-branch --branch '${GIT_TAG}' --quiet '${REPO_UNDER_TEST}' > /dev/null"
```
Better yet, support disabling this for debugging.
```bash
perform "git clone --recursive --single-branch --branch$([[ "$DEBUG" != 'true' ]] || echo ' --quiet') '${GIT_TAG}' '${REPO_UNDER_TEST}'$([[ "$DEBUG" != 'true' ]] || echo ' > /dev/null')"
```
Not sure if both `--quiet` and `> /dev/null` are needed, but [this example](https://stackoverflow.com/a/39103526) used both and the man page explicitly states that `--quiet` stops `git` from printing progress to `STDERR`, with no mention of `STDOUT`.

Reply:
This doesn't seem like a reason to hold up a PR.